runtime/src/lib.rs

```diff
@@ -139,10 +147,99 @@ pub unsafe extern "C" fn cairo_native__libfunc__hades_permutation(
 /// Felt252 type used in cairo native runtime
 #[derive(Debug)]
 pub struct FeltDict {
-    pub inner: HashMap<[u8; 32], *mut c_void>,
+    pub mappings: HashMap<[u8; 32], usize>,
+
+    pub layout: Layout,
+    pub elements: *mut (),
+
+    pub dup_fn: Option<extern "C" fn(*mut c_void, *mut c_void)>,
+    pub drop_fn: Option<extern "C" fn(*mut c_void)>,
+
     pub count: u64,
+}
+
+impl Clone for FeltDict {
+    fn clone(&self) -> Self {
+        let mut new_dict = FeltDict {
+            mappings: HashMap::with_capacity(self.mappings.len()),
+
+            layout: self.layout,
+            elements: if self.mappings.is_empty() {
+                null_mut()
+            } else {
+                unsafe {
+                    alloc(Layout::from_size_align_unchecked(
+                        self.layout.pad_to_align().size() * self.mappings.len(),
+                        self.layout.align(),
+                    ))
+                    .cast()
+                }
+            },
+
+            dup_fn: self.dup_fn,
+            drop_fn: self.drop_fn,
+
+            // TODO: Check if `0` is fine or otherwise we should copy the value from `old_dict` too.
```

Comment:
Yes, I just tried both and both passed the tests. I think it may be irrelevant because the way to clone a dict is to take a snapshot, but we cannot do anything with that snapshot so it remains unused.

Should I change the todo into a note?